### DISCUSSION TITLE
Dynamic Quantization in OpenVINO

### DISCUSSION THREAD

**Discussion by Nikitha-Shreyaa** [openvinotoolkit/openvino#25075] [Q&A]
I am trying to perform inference for a few LLM models using OpenVINO. My machine by default supports bfloat16. When I checked for the logs when performing inference for the models, the inference is happened to be at bfloat16. To perform the inference at fp32 I used "INFERENCE_PRECISION_HINT":"f32". 

Now I am trying to perform dynamic quantization for "distilbert-base-uncased-finetuned-sst-2-english" model. If I wanted to do so, after getting the int8 weight compressed model, should I load the model as

```
model = OVModelForSequenceClassification.from_pretrained(
	model_path,
	ov_config={"DYNAMIC_QUANTIZATION_GROUP_SIZE": "32", "PERFORMANCE_HINT": "LATENCY"},
)  
```

or

```
model = OVModelForSequenceClassification.from_pretrained(
	model_path,
	ov_config={"INFERENCE_PRECISION_HINT":"f32", 
                           "DYNAMIC_QUANTIZATION_GROUP_SIZE": "32", "PERFORMANCE_HINT": "LATENCY"},
)  

```
Because when I checked the logs for both the above cases, 

in **case:1** I was not able to view anything related to dynamic quant but 
in **case:2** I got logs related to dynamic quant. The below is the part of the log where my doubt arises 

**by setting "INFERENCE_PRECISION_HINT":"f32"**
onednn_verbose,primitive,exec,cpu,inner_product,brgemm:avx512_core_vnni,forward_inference,src_f32::blocked:ab::f0 wei_u8:a:blocked:AB4b32a4b::f0 bia_f32::blocked:a::f0 dst_f32::blocked:ab::f0,attr-scratchpad:user attr-scales:wei:1 attr-zero-points:wei:1 **src_dyn_quant_group_size:32**;,,mb6ic768oc768,0.0251465

**Without setting "INFERENCE_PRECISION_HINT":"f32"**
onednn_verbose,primitive,exec,cpu,inner_product,brgemm:avx512_core_bf16,forward_inference,src_bf16::blocked:ab::f0 wei_u8:a:blocked:AB16b64a::f0 bia_bf16::blocked:a::f0 dst_bf16::blocked:ab::f0,attr-scratchpad:user attr-scales:wei:1 attr-zero-points:wei:1 ,,mb6ic768oc768,0.0180664

Does quantization has to be done from fp32 only and not from bfloat16?

[code.txt](https://github.com/user-attachments/files/15872170/code.txt)

**Comment by Ganesh-403** [comment]
@Nikitha-Shreyaa Your logs are actually consistent with how OpenVINO's dynamic quantization is designed.

The important point is that **dynamic quantization and `INFERENCE_PRECISION_HINT` are two separate mechanisms**:

- `INFERENCE_PRECISION_HINT` controls the **floating-point precision** used for inference (`f32`, `bf16`, `f16`).
- `DYNAMIC_QUANTIZATION_GROUP_SIZE` enables **dynamic activation quantization** for supported operations (e.g. MatMul/InnerProduct), while your model weights remain INT8. :contentReference[oaicite:0]{index=0}

The key observation is that **dynamic quantization is not enabled for BF16 inference**. This is explicitly documented in the Optimum Intel/OpenVINO documentation:

> *"As of release 2024.3, dynamic quantization is not enabled for BF16 inference."* :contentReference[oaicite:1]{index=1}

That explains the difference in your logs:

With

```python
ov_config={
    "INFERENCE_PRECISION_HINT": "f32",
    "DYNAMIC_QUANTIZATION_GROUP_SIZE": "32"
}
```

oneDNN reports:

```
src_f32
wei_u8
src_dyn_quant_group_size:32
```

which indicates that dynamic activation quantization is active.

Whereas without `INFERENCE_PRECISION_HINT`, your CPU defaults to BF16:

```
src_bf16
wei_u8
```

and there is no `src_dyn_quant_group_size` field because dynamic quantization is disabled for BF16 execution. :contentReference[oaicite:2]{index=2}

So to answer your question:

> **Does quantization have to be done from FP32 only and not from BF16?**

Not exactly.

The **weight-compressed INT8 model** itself is independent of the runtime precision. What changes is the **runtime optimization**:

- **FP32 inference** → dynamic activation quantization can be applied.
- **BF16 inference** → currently, OpenVINO does not apply dynamic quantization, so you'll only see BF16 activations with INT8 weights. :contentReference[oaicite:3]{index=3}

Your logs therefore look like the expected behavior rather than a configuration issue.

If this solves your problem, feel free to mark it as the accepted answer so others can find it easily.
